File: content/11-graphs/course_schedule.js
function canFinish(numCourses, prerequisites) {
    let preMap = {};
    for (let i = 0; i < numCourses; i++) preMap[i] = [];
    for (let [crs, pre] of prerequisites) preMap[crs].push(pre);

    let visiting = new Set();

    function dfs(crs) {
        if (visiting.has(crs)) return false;
        if (preMap[crs].length === 0) return true;
        visiting.add(crs);
        for (let pre of preMap[crs]) {
            if (!dfs(pre)) return false;
        }
        visiting.delete(crs);
        preMap[crs] = [];
        return true;
    }

    for (let c = 0; c < numCourses; c++) {
        if (!dfs(c)) return false;
    }
    return true;
}

const assert = require('assert');
assert.strictEqual(canFinish(2, [[1,0]]), true);
assert.strictEqual(canFinish(2, [[1,0],[0,1]]), false);
console.log("All tests passed!");

File: content/11-graphs/course_schedule.py
def canFinish(numCourses: int, prerequisites) -> bool:
    preMap = {i: [] for i in range(numCourses)}
    for crs, pre in prerequisites:
        preMap[crs].append(pre)

    visiting = set()

    def dfs(crs):
        if crs in visiting:
            return False
        if preMap[crs] == []:
            return True
        visiting.add(crs)
        for pre in preMap[crs]:
            if not dfs(pre):
                return False
        visiting.remove(crs)
        preMap[crs] = []
        return True

    for c in range(numCourses):
        if not dfs(c):
            return False
    return True

if __name__ == "__main__":
    assert canFinish(2, [[1,0]]) == True, "Test 1 Failed"
    assert canFinish(2, [[1,0],[0,1]]) == False, "Test 2 Failed"
    print("All tests passed!")
